src/main/java/org/variantsync/diffdetective/variationtree/VariationTreeNode.java

```diff
@@ -0,0 +1,364 @@
+                new ArrayList<>()
+        );
+    }
+
+    /**
+     * Convenience constructor for creating an artifact node of an {@link VariationTree}.
+     *
+     * @param fromLine the start line number of the code of {@code label}
+     * @param toLine the end line number (exclusive) of the code of {@code label}
+     * @param label a list of lines used as label
+     * @see addBelow
+     */
+    public static VariationTreeNode createArtifact(int fromLine, int toLine, List<String> label) {
+        return new VariationTreeNode(NodeType.ARTIFACT, fromLine, toLine, null, label);
+    }
+
+    @Override
+    public VariationTreeNode getBackingNode() {
+        return this;
+    }
+
+    @Override
+    public NodeType getNodeType() {
+        return nodeType;
+    }
+
+    @Override
+    public List<String> getLabelLines() {
+        return label;
+    }
+
+    /**
+     * Replaces the current label by {@code newLabelLines}.
+     *
+     * <p>The given list is not copied, so modifications of {@code newLabelLines} will be visible by
+     * {@link getLabelLines}.
+     *
+     * @see getLabelLines
+     */
+    public void setLabelLines(List<String> newLabelLines) {
+        label = newLabelLines;
+    }
+
+    /**
+     * Adds the given lines to the source code lines of this node.
+     *
+     * @param lines lines to add
+     * @see getLabelLines
+     */
+    public void addLines(final List<String> lines) {
+        this.label.addAll(lines);
+    }
+
+    @Override
+    public int getFromLine() {
+        return from;
+    }
+
+    @Override
+    public void setFromLine(int from) {
+        this.from = from;
+    }
+
+    @Override
+    public int getToLine() {
+        return to;
+    }
+
+    @Override
+    public void setToLine(int to) {
+        this.to = to;
+    }
+
+    @Override
+    public VariationNode<VariationTreeNode> getParent() {
+        return parent;
+    }
+
+    /**
+     * Sets {@code newParent} as the new parent of this node.
+     *
+     * <p>This node must not have a parent yet.
+     */
+    private void setParent(final VariationTreeNode newParent) {
+        Assert.assertTrue(parent == null);
+        this.parent = newParent;
+    }
+
+    @Override
+    public List<VariationNode<VariationTreeNode>> getChildren() {
+        return Collections.unmodifiableList(childOrder);
+    }
+
+    @Override
+    public void addChild(final VariationNode<VariationTreeNode> child) {
+        child.getBackingNode().setParent(this);
+        childOrder.add(child.getBackingNode());
+    }
+
+    @Override
+    public void addChild(final VariationNode<VariationTreeNode> child, int index) {
+        child.getBackingNode().setParent(this);
+        childOrder.add(index, child.getBackingNode());
+    }
+
+    @Override
+    public boolean removeChild(final VariationNode<VariationTreeNode> child) {
+        if (isChild(child)) {
+            child.getBackingNode().parent = null;
+            childOrder.remove(child);
```

Comment:
`childOrder.remove(child.getBackingNode())`? Why doesn't Java complain here? I guess because the list methods just require `Object`s for some reason. 🙄

Reply:
Good catch.